### DISCUSSION TITLE
Using dyn4j for network layout

### DISCUSSION THREAD

**Discussion by wbpaley** [dyn4j/dyn4j#234] [Show and tell]
If anyone's interested, I'm exploring new techniques for laying out networks--_human_-directed techniques, not bot-only layered/hierarchical (Sugiyama) or force-directed layout. (Thesis: humans are better at putting meaning into graphs than algorithms are. The latter can minimize, e.g., crossings--but are intentionally semantics-free or else they wouldn't be general. Of course the goal is to have humans put in the meaning they want to and have bots do the housekeeping--dyn4j is a huge help in letting me create that housekeeping layer!)

So I let people drag nodes around to help them understand tangled networks. (You wouldn't untie a shoelace-knot with statistics, or with your hands tied behind your back! A large part of thinking can be called "thinking with your hands.")

https://drive.google.com/file/d/18C0PxfLGjVISXRA3JUqSnaoGR8gww4XD/view?usp=sharing

This quick demo builds an example network by letting Java Flight Recorder sample the stack of the running Java program (our analytics environment "the Pool"), making nodes from the classes, links and messages from the calls (adjacent layers in the stack). The final colorway also has a semantic mapping: it refers to blood oxygenation, letting one quickly see the relationship of one-hop nodes to the selected ones: blue nodes come before, red ones after, and the selected ones are white. Nodes that are both before and after are purple (though let's hope your own circulatory system doesn't have such loops!..  ; )

**Comment by wnbittle** [comment]
Wow, that super cool and looks great!  Could you describe what features were used?  Would love to hear if you have any constructive feedback on dyn4j (good or bad).  I'm also interested in the UI you built, was that custom, Swing, Java FX, or?

**Reply by wbpaley** [reply]
Funny: I replied to the email I got from notifications@github.com a while ago, thinking it might be forwarded to you and show up here--but no response & no post. GitHub might want to think about that. No pressure if you just didn't have anything to say, but I didn't want to appear unfriendly if you never got it, so here I cut & paste. (Still having great fun with it as my other show&tell demonstrates; thanks!)

---

Pardon my slow reply; trying to formulate a useful, complete, concise response.

Thanks for the praise!

I'm using just the bare minimum collision detection & non-overlapping Convex (Polygon & Ellipse) constraints for now. I get the margins by expanding the node/group geometry before handing it to dyn4j--basically I feed your physics world with larger objects hidden "underneath" the ones I draw. Works very nicely, as you see. But one neat extension might be the ability to define a margin around fixtures--I bet it'd be more efficient to solve that at your physical level. And maybe introduce things like soft margins--essentially area springs normal to the "almost-contact" point(s). Starting to look at using Joints as springs to maybe let you pull one node & have others follow it, or make stronger links enforce a minimum distance.

Have had many ideas about how I'll need to conceptually extend it--but they may or may not be useful to you or your audience, since I'm operating at a semantic layout level as well the physical "don't overlap" level. For instance I have a need to make a list of subnodes "look like a list" when, e.g., one opens up one of my classes (the elliptical nodes) to see the methods inside. Typographic convention has made lists simple stacks aligned in y for millennia, so I need a constraint that will align them in y. Also, I may want downstream nodes to be below upstream ones (in the video message flow is from top to bottom)--I realize this means the general graph will entail circular dependencies--that's part of the interesting challenge both at the physical and semantic levels. (I bet there are ways to use what you've already implemented to do at least part of this.)

I think these extensions should live in a separate package that uses dynj4 concepts/objects/solvers, but doesn't clutter up the careful, physics-targeted structure of dyn4j--maybe sem4j?

Have never found a UI or Visualization library or platform that didn't limit me more than it helped. Started our combined infrastructure VisUI 20 years ago, and are now creating the object-oriented coding environment on top of it--where "objects" are physical-looking entities on the screen and "coding" is generally the act of wiring them together. Could be called as visual programming language--but since almost none of those have ever succeeded I avoid the term; also, our environment does both more and less than VPLs try to do. Long philosophical discussion--terminating it here...

Afraid this email is neither complete nor concise--though I hope it's a bit useful. I've found that conversations are many times more productive than emails for this kind of collision/overlap of conceptual worlds.

Love to give you a demo (the video shows maybe 1/10 the functionality) & brainstorm if you're interested!
